### ISSUE TITLE
[BUG] Memory allocation issue whenever number of voxels increases a certain threshold.

### ISSUE THREAD

**Issue by k1moradi** [AcademySoftwareFoundation/openvdb#1489]
### Environment
**Operating System: Ubuntu 20.04.5 via nix-env
**Version / Commit SHA: 9.1 found in nux-env unstable channel
**Other: GCC 11.2, cmake 3.21

### Describe the bug

I'm collaborating on the development of an open-source program ([recut](https://github.com/UCLA-VAST/recut)) that is a program to reconstruct neurons using terabyte-size 3D images of the mouse brain. We use OpenVDB and hit a wall at the moment. Whenever the number of voxel sizes increases more than a certain number, some OpenVDB functions start to find trouble allocating memory. This problem happens in spite of the fact that our computer has 4TB of RAM and our runs barely hit the 1TB ram usage.

The following image shows the gdb output of one of the runs in which our program froze reaching `segmentActiveVoxels` function of OpenVDB.
![image](https://user-images.githubusercontent.com/18602635/195468840-e76bd32c-2331-4ebb-ace4-099f16659c8a.png)

I'm not a gdb wizard by to my understanding this is a memory allocation issue.

### To Reproduce
I might be able to produce a test code as needed.

### Expected behavior
`segmentActiveVoxels` works fine whenever the number of voxels is smaller than ~700,000,000. The program froze when the voxel numbers were more than ~800,000,000.

We expected voxel numbers smaller than 2^32 to work at least, although we can definitely use more voxel numbers if that works.

### Additional context
This is a very important university project. :) 
The testing hardware is a dual-socket HGX server equipped with AMD Epic CPUs.

**Comment by Idclip** [comment]
Hi @k1moradi,

Unfortnately there's not enough information in that backtrace to diagnose this or even confirm it is a memory allocation issue. I assume we don't want to run your 1TB tests in debug mode, so what we can try doing is adding more debug info to the release builds to attempt to get some more information from the trace. Could you try building your software with: `-g -ggdb3` ? That may give us a start to the general area in the tool header that's causing the issue. If that doesn't help you can also try rebuilding VDB with those flags (pass cmake `-DCMAKE_CXX_FLAGS="-g -ggdb3").

Let me know if you're able to get any more details from that.

**Comment by k1moradi** [comment]
I compiled recut software with the suggested flags.
![image](https://user-images.githubusercontent.com/18602635/196561877-908af594-0f42-4f00-8011-2b1257ddc279.png)
Do you understand anything more from these outputs? I'm not a gdb expert but to me, this output is similar to the previous output.

**Comment by Idclip** [comment]
Yes there's unfortunately no new information here - you'll also need to rebuild VDB with these flags too.

**Comment by k1moradi** [comment]
[An example vdb file](https://drive.google.com/file/d/1VXFPD452hm0jb72qO82Heq8NmjPWcKk7/view?usp=sharing).
We use the segmentSDF function like this:
```cpp
openvdb::io::File file("/path/to/file");
file.open();
openvdb::GridBase::Ptr sdf_grid= file.readGrid(file.beginName().gridName());
file.close();
std::vector<openvdb::FloatGrid::Ptr> components;
openvdb::v9_1::tools::segmentSDF(*sdf_grid, components);
```
